src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,561 @@
-        index_name = intersection_matrix.index.name
+    Break two layers, each covering the same area, into disjoint geometries.
 
-    else:
-        unique_index = intersection_matrix.columns
-        index_name = intersection_matrix.columns.name
+    Two GeoDataFrames are combined together in such a fashion that the
+    geometries are completely disjoint. The uniform attributes are allocated on
+    the basis of the fraction of the area covered by the new geometry
+    compared to the geometry it is being split from. There may be non-unique
+    geometries involved in either layer. If non-unique geometries are involved
+    in layer 1, layer 2 attributes get counted multiple times and are scaled
+    down accordingly and vice-versa.
 
-    if normalize is True:
-        intersection_matrix = intersection_matrix.divide(
-            intersection_matrix.sum(axis=axis_scale), axis=(1 - axis_scale))
+    Example:
+        In the case of simple geometries A, B and A intersection B (X2)
+        in layer 1, and layer 2 containing geometries 1 and 2. The
+        new geometry (1 int A int B) will be counted twice, and same for new
+        geometry (2 int A int B). However, the allocation of the uniform
+        attribute is done based on the area fraction. So, it is divided by the
+        number of times the duplication is occurring.
 
-    return intersection_matrix.multiply(
-        gdf_scale[gdf_scale[index_name].isin(unique_index)]
-        .set_index(index_name)[gdf_scale_col], axis=(1 - axis_scale))
+    The function returns a new GeoDataFrame with all columns from layer1 and
+    layer2.
 
+    Args:
+        layer1 (GeoDataframe): first GeoDataFrame
+        layer2 (GeoDataframe): second GeoDataFrame
+        attributes (dict): a dictionary keeping a track of all the types of
+        attributes with keys represented by column names from layer1 and
+        layer2, and the values representing the type of attribute. Types of
+        attributes include "constant", "uniform" and "ID". If a column name
+        `col` of type "ID" exists, then one column name `col`+"_set" of type
+        "constant" will exist in the attributes dictionary.
 
-def extract_multiple_tracts_demand_ratios(pop_norm_df, intermediate_ids):
+    Returns:
+        GeoDataFrame: New layer consisting all attributes in layer1 and layer2
+    """
+    # separating the uniforms and constant attributes
+    layer1_uniforms = [k for k, v in attributes.items() if (
+        (k in layer1.columns) and (v == "uniform"))]
+    layer2_uniforms = [k for k, v in attributes.items() if (
+        (k in layer2.columns) and (v == "uniform"))]
+
+    layer1_constants = [k for k, v in attributes.items() if (
+        (k in layer1.columns) and (v != "uniform"))]
+    layer2_constants = [k for k, v in attributes.items() if (
+        (k in layer2.columns) and (v != "uniform"))]
+
+    # Calculating the intersection layers
+    layer_new = geopandas.overlay(layer1, layer2)
```

Comment:
By default, `geopandas.overlay()` only returns the geometries of overlap between layers (`how='intersection'`). Unless Catalyst has no plans to use rows with incomplete data for demand modeling, this should be configurable, and the default should be `'union'`, since it preserves all geometries. This will require modifying the `merge()` below (and possibly other lines).

Reply:
Aah, yes. Right! Thank you. Although, I am slightly unsure what to do if the layers do not completely overlap. What should change?

Reply:
`geopandas.overlay(..., how='union')` will return all disjoint geometries, some of which only belong to one or the other layer, rather than an intersection between the layers. Those single-layer geometries will only have values for its attributes from its layer and missing values for attributes from the other layer.

Reply:
Yes, that's what I am worried about. Right now, I am considering that the geometries will be near perfect overlaps, as we are just focusing on the entire US mainland. I haven't figured out what to do if there are significant geometries which do not overlap. Can I just leave this in the current form for now and keep this flaw open as a separate issue?

Reply:
@yashkumar1803 Nothing inherently wrong with missing values in attributes – they would rightly represent that the layer with that attribute is not present at that location. I'll leave it to @zaneselvans to comment on the priority of propagating areas with incomplete attribute records to the demand modeling stage.